test/azul_test_case.py

```diff
@@ -1,10 +1,49 @@
-import os
-from unittest import TestCase, mock
+from unittest import TestCase
 
 import boto3.session
 
 
-class AzulTestCase(TestCase):
+class SafeTestCase(TestCase):
+    """
+    SafeTestCase makes sure that tearDown / cleanup methods are always run when
+    They should be.
+
+    This means that
+    - if a KeyboardInterrupt is raised in a test, then tearDown, tearDownClass,
+      and tearDownModule will all still run
+    - if any exception is raised in a setUp, then tearDown, tearDownClass,
+      and tearDownModule will all still run
+
+    Caveats:
+    - All tearDown methods MUST be idempotent. It's possible a step of tearDown
```

Comment:
I don't think that's accurate. They only need to handle the case that the corresponding setUp() hasn't finished. An idempotent tearDown is likely to have that property but I'm not sure they are equivalent.

Reply:
Ahh, I see. It's a subtle distinction, but I think yes, you're correct.